### PR TITLE
Allow user to edit a Blood Sugar within 60 minutes of it being added - Part 2

### PR DESCRIPTION
**Pull request by msasikanth**
https://www.pivotaltracker.com/story/show/169676402

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUi.kt [hunk 1]
```diff
@@ -18,5 +18,8 @@ interface BloodSugarEntryUi {
   fun hideDateErrorMessage()
   fun setDateOnInputFields(dayOfMonth: String, month: String, twoDigitYear: String)
   fun showDateOnDateButton(date: LocalDate)
+  fun showEditRadomBloodSugarTitle()
+  fun showEditPostPrandialBloodSugarTitle()
+  fun showEditFastingBloodSugarTitle()
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUi.kt]: Can we rename the corresponding methods for setting the title when entering a new blood sugar reading as well?

Example:
`showRandomBloodSugarTitle` -> `showEnterRandomBloodSugarTitle`.

File: app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRendererTest.kt [hunk 2]
```diff
@@ -58,4 +59,46 @@ class BloodSugarEntryUiRendererTest {
...
+    // given
+    val testUserClock = TestUserClock()
+    val year = LocalDate.now(testUserClock).year
+    val bloodSugarEntryModel = BloodSugarEntryModel.create(year, Update(bloodSugarUuid, Fasting))
+
+    // when
+    bloodSugarEntryUiRenderer.render(bloodSugarEntryModel)
+
+    // then
+    verify(ui).showEditFastingBloodSugarTitle()
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRendererTest.kt]: Should we also add `verifyNoMoreInteractions(ui)` after the assertions in these methods?

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRenderer.kt [hunk 3]
```diff
@@ -35,6 +35,7 @@ class BloodSugarEntryUiRenderer(
 
   private fun showEditBloodSugarTitle(measurementType: BloodSugarMeasurementType) {
     with(ui) {
+      showRemoveButton()
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRenderer.kt]: The name of this method is `showEditBloodSugarTitle`, but it also does showing of the remove button. Given that the showing of the remove button is not conditionally triggered based on the `measurementType`, it seems to me that this method call here is because of convenience and not because it actually belongs here.

What do you think about moving this into the `render()` method instead? We can just decide to show/hide this based on the `openAs` property.

File: app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRendererTest.kt [hunk 4]
```diff
@@ -85,6 +86,7 @@ class BloodSugarEntryUiRendererTest {
     bloodSugarEntryUiRenderer.render(bloodSugarEntryModel)
 
     // then
+    verify(ui).showRemoveButton()
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRendererTest.kt]: Should we add separate tests for this?

`When the sheet is opened for editing a blood sugar measurement, the remove button must be shown`, and `When the sheet is opened for recording a new blood sugar measurement, the remove button must not be shown`

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRenderer.kt [hunk 5]
```diff
@@ -25,6 +25,7 @@ class BloodSugarEntryUiRenderer(
 
   private fun showEnterNewBloodSugarTitle(measurementType: BloodSugarMeasurementType) {
     with(ui) {
+      hideRemoveButton()
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRenderer.kt]: Same comment as earlier about the method being here because of convenience, and moving it into the `render` function.

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUpdate.kt [hunk 6]
```diff
@@ -33,7 +32,7 @@ class BloodSugarEntryUpdate @AssistedInject constructor(
     return when (event) {
       is ScreenChanged -> next(model.screenChanged(event.type))
       is DatePrefilled -> next(model.datePrefilled(event.prefilledDate))
-      is BloodSugarMeasurementFetched -> noChange()
+      is BloodSugarMeasurementFetched -> onBloodSugarMeasurementFetched(model, event)
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUpdate.kt]: The prefix `on` indicates that this is some form of callback method, while this is a method that returns a `Next`. I feel we can just drop the `on` prefix.

What do you think?

File: app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryEffectHandlerTest.kt [hunk 7]
```diff
@@ -162,6 +162,22 @@ class BloodSugarEntryEffectHandlerTest {
     verifyNoMoreInteractions(ui)
   }
 
+  @Test
+  fun `set specific date on input fields and date button when prefill date for update effect is received`() {
+    // given
+    val entryDate = LocalDate.of(2020, 2, 14)
+    userClock.setDate(LocalDate.of(2020, 2, 14))
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryEffectHandlerTest.kt]: Can we reuse `entryDate` for setting the date for `userClock`?

File: app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryEffectHandlerTest.kt [hunk 8]
```diff
@@ -162,6 +162,22 @@ class BloodSugarEntryEffectHandlerTest {
...
+
+    // when
+    testCase.dispatch(PrefillDate.forUpdateEntry(userClock.instant()))
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryEffectHandlerTest.kt]: We generally use `Instant.now(clock)` instead of `clock.instant()` throughout the codebase. Could we follow this convention here as well, please?

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntrySheet.kt [hunk 9]
```diff
@@ -324,6 +324,12 @@ class BloodSugarEntrySheet : BottomSheetActivity(), BloodSugarEntryUi {
     finish()
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntrySheet.kt]: Was "blood" in the commit message meant to be "blood sugar"?

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryInit.kt [hunk 10]
```diff
@@ -7,8 +7,8 @@ import com.spotify.mobius.Init
 class BloodSugarEntryInit : Init<BloodSugarEntryModel, BloodSugarEntryEffect> {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryInit.kt]: Can we add a line to the commit message as to why this change was made?

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUi.kt (hunk 1 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUi.kt]: In later commits, I am planning on replacing these methods with a single method with just the measurement type parameter. Should I still replace these methods names in this PR?

File: app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRendererTest.kt (hunk 2 above)
Comment (msasikanth) [reply app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRendererTest.kt]: Yes, forgot about that

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRenderer.kt (hunk 3 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRenderer.kt]: Moved the `showRemoveButton` inside the `setupUi` method

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRenderer.kt (hunk 5 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRenderer.kt]: Done 👍

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUpdate.kt (hunk 6 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUpdate.kt]: Yeah 👍

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntrySheet.kt (hunk 9 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntrySheet.kt]: Yes, reworded the commit message `Disable background click of sheet if blood sugar reading is not blank`

File: app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUi.kt (hunk 1 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUi.kt]: If you're going to do this, then it's fine. Let's just take a look at it in the end after the feature is complete.

File: app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRendererTest.kt (hunk 4 above)
Comment (vinaysshenoy) [reply app/src/test/java/org/simple/clinic/bloodsugar/entry/BloodSugarEntryUiRendererTest.kt]: Discussed offline: the tests are already present and were written earlier.
